php/7.4/Dockerfile

```diff
@@ -3,6 +3,18 @@ FROM phpdockerio/php74-fpm:latest
 # Fix debconf warnings upon build
 ARG DEBIAN_FRONTEND=noninteractive
 
+# Passed via an environment variable to docker-compose.yml build: args:
+ARG SQ1_UID=1000
```

Comment:
I can confirm that this does not break Docker on my Mac when using Mutagen caching (haven't tested it yet with `osxfs`). But I don't see how this arg gets passed to the build that will run on Docker Hub. Our local systems will just be pulling the image, not rebuilding it.

Reply:
Right...that ARG will only run on build. Hrmmm then the only way would be to include a Dockerfile in square one? I started a PR for this here: https://github.com/moderntribe/square-one/pull/493

Reply:
Or perhaps this can be a `docker-entrypoint.sh` script?

Reply:
I'm mostly ignorant of the details of file permissions with Docker on Linux, so forgive me if this is nonsense...

Can you set the `user` on the container, as Luca suggested for tric, in a `docker-compose.override.yml`? Or does that cause other issues?

Is it a requirement that we create a new user? Can you also use `docker-compose.override.yml` to mount another config file at `/etc/php/7.4/fpm/pool.d/zzz-overrides.conf`?

Reply:
I'm glad you asked so I can document the exact issues we want to solve:

1. The default user that executes anything on this container is root. Because docker uses the native Linux kernel, anything that happens in the container translates up to the host.
2. On this particular image, the default fpm pool is running as the `www-data` user. 
3. When we run docker exec, for example `composer install` it will execute as root and everything installed by composer, cache dir, plugins, WordPress etc...are now owned by root both in the container AND on the host. So I have to chmod my `composer.lock` file, for example of I want to delete it. This can be solved by passing the `--user`  flag to the docker-compose exec command.
4. As fpm actions happen we run into more problems. For example, loading a site that doesn't have a twig cache folder will often not be able to create that folder because all the parent folders are own by root or by my host user. Same goes for the uploads folder. Making the fpm pool run as the same user as the host solves this so every time fpm has to write something, it'll have permission and so will I, say if I wanted to manually delete something from the uploads folder without having to use sudo or chmod'ing it first.

Now, we don't want to tell this container to run everything as the host user because the main fpm process runs as root and needs write access to many things in the container, like /run, and log directories and such. I did try it and it the container fails to come up as it's the fpm process is unable to write the pid file to /run.

Mounting the custom config file is not enough as a user matching the hosts UID/GID has to exist in the container in order to tell the pool which user run the pool as unless it will take id's (which I didn't test, but I've also never seen before).

If you have any other ideas let me know. I did test this on the PlanetWare project by modifying its dockerfile and it works quite well, but I'd like a solution for everyone on Linux in place.

Reply:
I'm going to think out loud here about how we could solve this with an entrypoint script.

The script could check for environment variables for HOST_UID and HOST_GID. If they're set, it creates a new `squareone` user and group in the container. It then creates a file at `/etc/php/7.4/fpm/pool.d/zzz-overrides.conf` with just:

```
[www]
; a custom user that matches the host's user ID and created in the Dockerfile
user = squareone
group = squareone
```

Alternatively, we could give the `www-data` user a different user ID/group ID.

After that, the entrypoint is a simple passthrough:

```
exec "$@"
```

... But this only solves PHP-FPM, not other CLI actions/docker exec/composer. Those really require the container to run as another user. I want to prod more at why the entire container can't run as a different user. Perhaps we should have a meeting to walk through it.

Reply:
I'm already working on it, and I can make exec commands run as the user properly in my newer version of squareone global docker.